Always validate checksums for Direct I/O reads

bwatkinson opened the pull request:
This fixes an oversight in the Direct I/O PR. There is nothing stops a process from manipulating the contents of a buffer for a Direct I/O read while the I/O is in flight. This can lead checksum verify failures. However, the disk contents are stil correct, but this would lead false reporting of checksum validations.

To remedy this, all Direct I/O reads that have a checksum verification failure are treated as suspicious. In the event a checksum validation failure occurs for a Direct I/O read, then the I/O request will be reissued though the ARC. This allows for actual validation to happen and removes any possibility of the buffer being manipulated after the I/O has been issued.

Just as with Direct I/O write checksum validation failures, Direct I/O read checksum validation failures are reported though zpool status -d in the DIO columnm. Also the zevent has been updated to have both:
1. dio_verify_wr -> Checksum verification failure for writes
2. dio_verify_rd -> Checksum verification failure for reads. This allows for determining what I/O operation was the culprit for the checksum verification failure. All DIO errors are reported only on the top-level VDEV.

Even though FreeBSD can write protect pages (stable pages) it still has the same issue as Linux in department.

This commit updates the following:
1. Propogates checksum failures for reads all the way up to the top-level VDEV.
2. Reports errors through zpool status -d as DIO.
3. Has two zevents for checksum verify errors with Direct I/O. One for read and one for write.
4. Updates FreeBSD ABD code to also check for ABD_FLAG_FROM_PAGES and handle ABD buffer contents validation the same as Linux.
5. Moves the declartion of nbytes in zfs_read() to the top of the function and outside of the while loop. This was needed due to a compliation failure in FreeBSD.
6. Updated manipulate_user_buffer.c to also manipulate a buffer while a Direct I/O read is taking place.
7. Adds a new test case dio_read_verify that stress tests the new code.

This issue was first observed when installing a Windows 11 VM on a ZFS dataset with the dataset property direct set to always. The zpool devices would report checksum failures, but running a subsequent zpool scrub would not repair any data and report no errors.

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->
Make sure that checksum verification failures with Direct I/O reads are reported and always reissued through the ARC.
<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Without this, there are false positive checksum verification failures reported when a Direct I/O read is issued and a process is manipulating the buffer contents.
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
<!--- Describe your changes in detail -->
See above.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
Added new test case. Tested on Linux: 6.5.12-100.fc37.x86_64 kernel. On that same Linux kernel I used Virtual Manger to install Windows 11, Ubuntu 22.04, and Fedora 38 on a dataset with the direct dataset property set to always. Ran ZTS direct test cases as well.

On FreeBSD tested ZTS direct tests on FreeBSD 13.3 -RELEASE and FreeBSD 14-CURRENT VM's.
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

tonyhutter commented on module/os/freebsd/zfs/abd_os.c:
```diff
@@ -632,8 +639,34 @@ abd_return_buf(abd_t *abd, void *buf, size_t n)
+#ifdef ZFS_DEBUG
+		/*
+		 * We have to be careful with gang ABD's that we do not ASSERT
+		 * for nay ABD's that contain user pages from Direct I/O. See
+		 * the comment above about Direct I/O read buffers possibly
+		 * being manipulated. In order to handle this, we jsut interate
```
"for any ABDs"
"we just iterate"

bwatkinson replied:
Updated.

tonyhutter commented on module/os/linux/zfs/abd_os.c:
```diff
@@ -1008,7 +1008,9 @@ abd_borrow_buf_copy(abd_t *abd, size_t n)
  * write protection and the user could have possibly changed the contents in
- * the pages at any time.
+ * the pages at any time. This is also an issue for Direct I/O reads. Checksum
+ * verifications in the ZIO pipeline check fo rthis issue and handle it by
```
"check for this issue"

bwatkinson replied:
Updated.

tonyhutter commented on tests/zfs-tests/tests/functional/direct/dio_read_verify.ksh:
```diff
@@ -0,0 +1,107 @@
+# STRATEGY:
+#	1. Create a zpool from each vdev type.
+#	2. Start a Direct I/O read workload while manipulating the user buffer
+#	   without using 
```
"without using.."?

bwatkinson replied:
Updated.

tonyhutter commented:
This should give you the JSON:
```diff
diff --git a/cmd/zpool/zpool_main.c b/cmd/zpool/zpool_main.c
index aa7da68aa..90f4225e1 100644
--- a/cmd/zpool/zpool_main.c
+++ b/cmd/zpool/zpool_main.c
@@ -9222,6 +9222,11 @@ vdev_stats_nvlist(zpool_handle_t *zhp, status_cbdata_t *cb, nvlist_t *nv,
                                fnvlist_add_string(vds, "power_state", "-");
                        }
                }
+               if (cb->cb_print_dio_verify) {
+                       nice_num_str_nvlist(vds, "dio_verify_errors",
+                           vs->vs_dio_verify_errors, cb->cb_literal,
+                           cb->cb_json_as_int, ZFS_NICENUM_1024);
+               }
        }
 
        if (nvlist_lookup_uint64(nv, ZPOOL_CONFIG_NOT_PRESENT,
```

behlendorf commented on module/zfs/zfs_vnops.c:
```diff
@@ -424,8 +427,26 @@ zfs_read(struct znode *zp, zfs_uio_t *uio, int ioflag, cred_t *cr)
+				 * suspicious. The contents of the buffer could
+				 * have beeen manipulated while the I/O was in
+				 * flight. In this case, the remainder of I/O
+				 * request will just be just be reissued through
```
```suggestion
				 * request will just be reissued through the ARC.
```

bwatkinson replied:
Updated.

behlendorf commented on module/zfs/zfs_vnops.c:
```diff
@@ -424,8 +427,26 @@ zfs_read(struct znode *zp, zfs_uio_t *uio, int ioflag, cred_t *cr)
+					uio->uio_extflg &= ~UIO_DIRECT;
+					n += dio_remaining_resid;
+					dio_remaining_resid = 0;
+					goto top;
```
The `goto` looks like it can be replaced with `continue` here and the "top" label removed.

bwatkinson replied:
Yeah, that makes a lot more sense. It also allowed me to put the `nbytes` declaration back into `zfs_read()`. I have updated this to use `continue` now.

behlendorf commented on module/zfs/zfs_vnops.c:
```diff
@@ -472,6 +493,10 @@ zfs_read(struct znode *zp, zfs_uio_t *uio, int ioflag, cred_t *cr)
 
+	if (dio_checksum_failure == B_TRUE) {
+		uio->uio_extflg |= UIO_DIRECT;
+	}
```
nit: parenthesis can be removed.

bwatkinson replied:
Updated.

behlendorf commented on module/zfs/zio.c:
```diff
@@ -1573,6 +1573,15 @@ zio_vdev_child_io(zio_t *pio, blkptr_t *bp, vdev_t *vd, uint64_t offset,
+		 * checksum verify failure is removed from the parent. If it
+		 * occurs again on the ditto block, then the parent will
+		 * require this flag for reporting the erorr to the top-level
+		 * Mirror VDEV in vdev_mirror_io_done().
```
```suggestion
		 * mirror VDEV in vdev_mirror_io_done().
```

bwatkinson replied:
Updated.

behlendorf commented on module/zfs/zio.c:
```diff
@@ -4864,16 +4873,38 @@ zio_checksum_verify(zio_t *zio)
+				 * Any Direct I/O read that has a checksum
+				 * error must be treated as suspicous as the
+				 * contents of the buffer could be getting
+				 * manipulated whiel the I/O is taking place.
```
```suggestion
				 * manipulated while the I/O is taking place.
```

bwatkinson replied:
Updated.

behlendorf commented on module/zfs/zio.c:
```diff
@@ -4864,16 +4873,38 @@ zio_checksum_verify(zio_t *zio)
+				 * The checksum verify error will only be
+				 * reported here for disk and file VDEV's and
+				 * will be reported on those that the failure
+				 * occurred on. Oter types of VDEV's report the
```
```suggestion
				 * occurred on. Other types of VDEV's report the
```

bwatkinson replied:
Updated.

behlendorf commented on module/zfs/zio.c:
```diff
@@ -4932,6 +4949,38 @@ zio_checksum_verified(zio_t *zio)
+void
+zio_dio_chksum_verify_error_report(zio_t *zio)
+{
+	ASSERT(zio->io_flags & ZIO_FLAG_DIO_CHKSUM_ERR);
```
nit: I'd add an empty line after the ASSET.

bwatkinson replied:
Updated.

behlendorf commented on module/zfs/zio.c:
```diff
@@ -4932,6 +4949,38 @@ zio_checksum_verified(zio_t *zio)
+	mutex_exit(&zio->io_vd->vdev_stat_lock);
+	if (zio->io_type == ZIO_TYPE_WRITE) {
+		/*
+		 * Convert checksum error error for writes into EIO.
```
```suggestion
		 * Convert checksum error for writes into EIO.
```

bwatkinson replied:
Updated.

behlendorf commented on tests/zfs-tests/tests/functional/direct/dio_read_verify.ksh:
```diff
@@ -0,0 +1,107 @@
+
+log_must  truncate -s $MINVDEVSIZE $DIO_VDEVS
+
+# We will verfiy that there is no checksum errors for every Direct I/O read
```
```suggestion
# We will verify that there are no checksum errors for every Direct I/O read
```

bwatkinson replied:
Updated.

behlendorf reviewed (approved):
The verification and error propagation looks good.  Thanks for running down this very strange but entirely possible case.  Most of the comments are for minor nits.

bwatkinson commented:
> This should give you the JSON:
> 
> ```diff
> diff --git a/cmd/zpool/zpool_main.c b/cmd/zpool/zpool_main.c
> index aa7da68aa..90f4225e1 100644
> --- a/cmd/zpool/zpool_main.c
> +++ b/cmd/zpool/zpool_main.c
> @@ -9222,6 +9222,11 @@ vdev_stats_nvlist(zpool_handle_t *zhp, status_cbdata_t *cb, nvlist_t *nv,
>                                 fnvlist_add_string(vds, "power_state", "-");
>                         }
>                 }
> +               if (cb->cb_print_dio_verify) {
> +                       nice_num_str_nvlist(vds, "dio_verify_errors",
> +                           vs->vs_dio_verify_errors, cb->cb_literal,
> +                           cb->cb_json_as_int, ZFS_NICENUM_1024);
> +               }
>         }
>  
>         if (nvlist_lookup_uint64(nv, ZPOOL_CONFIG_NOT_PRESENT,
> ```

Good catch. I add your patch to this PR.

bwatkinson commented:
@amotin and I had a conversation. It was incorrect for me to allow the mirror VDEV to issue reads down to other children in the event of a Direct I/O read checksum verify failure. The issue is, that another child could get a valid checksum, and then the buffer is manipulated again. This would lead to self healing in `vdev_mirror_io_done()` issue bad data down to the other children.

I updated `vdev_mirror_io_done()` to return immediately once a Direct I/O read checksum verification failure occurs. I also updated the comments in that code block. In `zio_vdev_child_io()` I now `ASSERT` that the parent can not have a Direct I/O checksum validation error. This makes sure children in the mirror are not being used to read after the initial checksum validation failure.

@behlendorf and @tonyhutter, can you relook at those two spots in the code and make sure my comments make sense to you?

behlendorf reviewed (commented):
Good catch on the `vdev_mirror_io_done()` case.  Yes, the updated logic looks correct to me, we always want to fall back to reading through the ARC in the event of a checksum failure.  It seems to me we also should do something similar for raidz and draid.  Repair IOs are issued from pages in the user ABD which could change after the checksum verification like with the mirror.

amotin commented on module/zfs/zio.c:
```diff
@@ -4864,16 +4872,38 @@ zio_checksum_verify(zio_t *zio)
-			zio->io_vd->vdev_stat.vs_checksum_errors++;
-			mutex_exit(&zio->io_vd->vdev_stat_lock);
-			(void) zfs_ereport_start_checksum(zio->io_spa,
-			    zio->io_vd, &zio->io_bookmark, zio,
-			    zio->io_offset, zio->io_size, &info);
+			if (zio->io_flags & ZIO_FLAG_DIO_READ) {
+				zio->io_flags |= ZIO_FLAG_DIO_CHKSUM_ERR;
```
It seems for speculative read (`ZIO_FLAG_SPECULATIVE`) you won't set `ZIO_FLAG_DIO_CHKSUM_ERR` here, that I guess may cause recovery writes on higher levels. It seems this `if` should move moved up.

bwatkinson replied:
I looked over the code, and I am not aware of anyway that a Direct I/O read could be issued as speculative. The only place I spotted as maybe a concern was data prefetching but that does not happen with Direct I/O reads. I did add an `IMPLY` statement to `zio_checksum_verify()` to make sure this is never the case though.

bwatkinson commented:
> Good catch on the `vdev_mirror_io_done()` case. Yes, the updated logic looks correct to me, we always want to fall back to reading through the ARC in the event of a checksum failure. It seems to me we also should do something similar for raidz and draid. Repair IOs are issued from pages in the user ABD which could change after the checksum verification like with the mirror.

Good catch. As per your suggestion, I now leverage the `rc_allow_repair` to make sure we never issue self healing for Direct I/O reads for raids and dRAID VDEVs.
